tests/dtypes_test.py

```diff
@@ -182,11 +182,11 @@ def testScalarCastInsideJitWorks(self):
 
 class TestPromotionTables(jtu.JaxTestCase):
 
-  def testWeakDtypesPromotionTable(self):
+  def testObservedPromotionTable(self):
```

Comment:
Delete this test? It's now redundant with the lattice test.

(Alternatively, change the code in dtypes.py to generate its table with the lattice, but don't repeat the table twice if it's a generated artifact.)

Reply:
This is a bit different: it's an integration test to confirm that the actual observed behavior of type promotion implemented in binary operators (typified by `jnp.add`) does not change.

The lattice test is an internal unit test that confirms the internal promotion table can be generated from a lattice.

For performance reasons, I suspect it's not a good idea to have `dtypes.py` generate the promotion table from the lattice, as that generation would happen at import time (although I've not benchmarked it).

Reply:
If the goal is to verify `jnp.add` has the behavior in the table, I suggest importing the type table from dtypes.py.

I'm mostly unhappy about the redundancy.

Reply:
JAX is already a pretty costly import all told, I'd rather avoid making it worse:
```python
%time import jax
# CPU times: user 395 ms, sys: 41.8 ms, total: 437 ms
# Wall time: 439 ms

%time import numpy
# CPU times: user 21 µs, sys: 0 ns, total: 21 µs
# Wall time: 24.6 µs
```

Reply:
The lattice->promotion table code currently takes ~1.5ms... so it's not much compared to JAX's full startup cost, but does contribute.

What do you think?

Reply:
We don't need a second "does not change" test, and for an end-to-end test, we don't need to repeat the table because we are testing the external behavior matches the table.

Reply:
The other option would be to use the lattice directly within `promote_dtypes`, possibly along with `lru_cache` to compute promotion rules on demand.

Reply:
> We don't need a second "does not change" test,

I'm not sure what the second test is that you're referring to.

> for an end-to-end test, we don't need to repeat the table because we are testing the external behavior matches the table.

What external behavior are you referring to?

Reply:
The lattice test also has the effect of a "does not change" test. You do not need two such tests.

For "external behavior" I mean what you refer to as "it's an integration test to confirm that the actual observed behavior of type promotion implemented in binary operators". You do not need to repeat the type promotion table to write such a test. There should only be one copy of the promotion table anywhere, and I'm happy for it to be the one in dtypes.py and for the test to merely check the actual behavior (as produced by operators) matches the expected behavior (as described by the table).

Reply:
> The lattice test also has the effect of a "does not change" test. You do not need two such tests.

I don't think that's quite true: the lattice test generates the internal table. The outputs of a binary operation are the internal table as interpreted by `result_type`, which had nontrivial logic prior to this PR. A big reason I created that "does not change" test was to ensure that the change of logic in `result_type` within this PR would lead to consistent results. For that reason, I think it's better not to remove it in the same commit that it's designed to validate; would you be open to me removing it in a followup?

> For "external behavior" I mean...

Ah, got it. There would need to be some non-trivial transformation logic, though, becuase the internal table includes weak types and the results of binary operations currently do not return weak types; also the behavior outside x64 mode differs. 

Because of the first point, I really think it would be better to keep this test for now, even if it's removed in a followup PR. What do you think?